### DISCUSSION TITLE
Magick++: Reading single image from multipage TIFF file - Performance

### DISCUSSION THREAD

**Discussion by uniqueusername2718** [ImageMagick/ImageMagick#7069] [Development] [answered]
### ImageMagick version

7.1.1

### Operating system, version and so on

Windows 11

### Description

I am reading a single page from a Blob containing a 65MB multipage TIFF file with 39 pages.  Each page is 2550 x 3300 x 24-bit. In my environment, the call to image.read(blob) in the code snippet below takes about 12 seconds. 
If I perform the same operation using a file containing only one image of the same size, image.read(blob) takes less than a second.
My suspicion is that the entire 39 page file is being decompressed in order to extract the desired page. My hope was that it would be possible to extract the single page directly without decompressing the whole file. Is it the case that it is decompressing the whole file? Are there settings that can be changed to improve this performance and/or change the behavior?

```
   Magick::Image image;
   image.subImage(page);
   try {
       Magick::Blob blob(buffer, bufferSize);
       image.read(blob);   // takes ~12s to read single page from 39 page file, <1s from 1 page file
   }
```
The 12s time is independent of the page selected (page 1 and page 35 take the same amount of time to extract.)

**Edit:** As a side note, reading from the file on disk using image.read("c:\\temp\\test.tif[3]") is fast. Is there a similar way to specify the page when reading from a Blob?

**Edit2:** I tried calling MagickCore::BlobToImage(imageInfo, buffer, bufferSize, exception) having set imageInfo->scenes to "3-3" and found that it also took 12 seconds. I'm not sure if this is correct format for "scenes"

**Comment by urban-warrior** [comment]
Try

```
Magick::Blob blob(buffer, bufferSize);
Magick::Image image(blob);
image.write("firstFrame.png");
```

That should pass the blob directly to the image container and only write the first image of the container.  Does that run in < 12s?

**Reply by uniqueusername2718** [reply]
No, Magick::Image image(blob) takes 12s.

I have found the following solution, which pulls out image/page 3 instantly:

```
      MagickCore::CloneString(&imageInfo->scenes, "3");
      imageInfo->number_scenes = 1;
      imageInfo->scene = 3;
      MagickCore::Image* image = BlobToImage(imageInfo, buffer, bufferSize, exception);
```

And, as mentioned above, I also note that `image.read("c:\temp\test.tif[3]")` pulls out image 3 instantly.

So, my questions now are:
a) Is there are way to implement the above solution using Magick++? There does not seem to be a way to set `number_scenes`.
b) Is my solution above, using MagickCore, the best way to do this?
c) And, if so, can the `MagickCore::Image` be easily converted to a `Magick::Image` (the rest of the app uses `Magick::Image`s)?

Thank you.

**Comment by uniqueusername2718** [comment, marked as answer]
Answering my own question, here is what I came up with to read, from a Blob containing a multipage TIFF, a particular page as a Magick::Image without decompressing the whole Blob. I am not experienced with ImageMagick, so I can't say this is the best way -- perhaps someone more experienced can chime in.

```
   MagickCore::Image* mcImage = nullptr;
   MagickCore::ImageInfo* imageInfo = nullptr;
   MagickCore::ExceptionInfo* exception = nullptr;
   try {
      imageInfo = MagickCore::CloneImageInfo(NULL);
      exception = MagickCore::AcquireExceptionInfo();

      // scenes are zero-based
      MagickCore::CloneString(&imageInfo->scenes, std::to_string(page - 1).c_str());
      imageInfo->number_scenes = 1;
      imageInfo->scene = page - 1;
      try {
         mcImage = BlobToImage(imageInfo, buffer, bufferSize, exception);
      }
      catch (const Magick::Warning& e) {
          GlobalLogging->WriteLog(LoggingManager::DEBUGLOG, e.what());
      }
      Magick::Image image(mcImage);
.
.  etc.
.
  }
```

**Comment by urban-warrior** [comment]
What about this Magick++ sequence?  Does that work for you?

```
 // Read the first image from the blob
Magick::Blob blob(blob_data.data(), blob_data.size());
Magick::Image image;
image.subRange(1);
image.read(blob);
```

**Reply by uniqueusername2718** [reply]
This is fast for reading the first image. What do I need to set to read an arbitrary image (e.g., page 3 or 36) from the blob? Thanks!

**Comment by urban-warrior** [comment]
Try `subImage()`:

```
 // Read the first image from the blob
Magick::Blob blob(blob_data.data(), blob_data.size());
Magick::Image image;
image.subImage(3);
image.subRange(1);
image.read(blob);
```

**Reply by uniqueusername2718** [reply]
The following code (same as above with image.write added) throws a Magick::ErrorCorruptImage exception  (NegativeOrZeroImageSize `myFrame.png' @ error/image.c/CloneImage/809) at the image.write statement. The image does indeed seem to have zero dimensions after the image.read(blob). I am using the same blob contents that works with the MagickCore core above.

```
Magick::Blob blob(blob_data.data(), blob_data.size());
Magick::Image image;
image.subImage(3);
image.subRange(1);
image.read(blob);
image.write("myFrame.png");
```

As a side note, I found with the MagickCore code that all three of the parameters below were required to get it work. Are there equivalent settings for Magick++?
```
      MagickCore::CloneString(&imageInfo->scenes, std::to_string(page - 1).c_str());
      imageInfo->number_scenes = 1;
      imageInfo->scene = page - 1;

```
